rest/backend/src/main/scala/io/udash/rest/server/ExposesREST.scala

```diff
@@ -0,0 +1,126 @@
+package io.udash.rest.server
+
+import javax.servlet.http.HttpServletRequest
+
+import com.avsystem.commons.rpc.MetadataAnnotation
+import io.udash.rest.{UdashRESTFramework, _}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+/**
+  * Base trait for anything that exposes REST interface.
+  */
+abstract class ExposesREST[ServerRPCType : UdashRESTFramework#ValidServerREST](localRest: ServerRPCType) {
+  val framework: UdashRESTFramework
+  import framework._
+
+  /**
+    * This allows the RPC implementation to be wrapped in raw RPC which will translate raw calls coming from network
+    * into calls on actual RPC implementation.
+    */
+  protected def localRpcAsRaw: AsRawRPC[ServerRPCType]
+
+  protected lazy val rawLocalRpc = localRpcAsRaw.asRaw(localRest)
+
+  protected val rpcMetadata: RPCMetadata[ServerRPCType]
+
+  /** Transform `String` from HTTP request header into `RawValue`. */
+  def headerArgumentToRaw(raw: String, isStringArg: Boolean): RawValue
+  /** Transform `String` from HTTP request query argument into `RawValue`. */
+  def queryArgumentToRaw(raw: String, isStringArg: Boolean): RawValue
+  /** Transform `String` from URL part into `RawValue`. */
+  def urlPartToRaw(raw: String, isStringArg: Boolean): RawValue
+
+  def handleRestCall(getterChain: List[RawInvocation], invocation: RawInvocation)(implicit ec: ExecutionContext): Future[String] = {
+    try {
+      val receiver = rawLocalRpc.resolveGetterChain(getterChain)
+      receiver.call(invocation.rpcName, invocation.argLists).map(rawToString)
+    } catch {
+      case ex: Exception =>
+        Future.failed(ex)
+    }
+  }
+
+  def parseHttpRequest(req: HttpServletRequest, httpMethod: Class[_ <: RESTMethod]): (List[RawInvocation], RawInvocation) = {
+    val invocations = List.newBuilder[RawInvocation]
+    val path: Array[String] = Option(req.getPathInfo).map(_.stripPrefix("/").split("/")).getOrElse(Array.empty[String])
+    val bodyContent = req.getReader.lines().toArray.mkString("\n")
+    lazy val bodyValues = read[Map[String, framework.RawValue]](stringToRaw(bodyContent))(bodyValuesReader)
+
+    def findRestParamName(annotations: Seq[MetadataAnnotation]): Option[String] =
+      annotations.find(_.isInstanceOf[RESTParamName]).map(_.asInstanceOf[RESTParamName].restName)
```

Comment:
Alternatively:
```scala
annotations.collectFirst {
  case rpn: RESTParamName => rpn.restName
}
```

Also, this method is only used like this, several times:
```scala
findRestParamName(arg.annotations).getOrElse(arg.name)
```
maybe it should just take `arg: ParamMetadata` and return `String`?